core/matching/matchers/xml_templated_match.go

```diff
@@ -0,0 +1,102 @@
+package matchers
+
+import (
+	"regexp"
+
+	"github.com/beevik/etree"
+)
+
+var XmlTemplated = "xmltemplated"
+
+var ignoreExpr = regexp.MustCompile("^\\s*\\{\\{\\s*ignore\\s*\\}\\}\\s*$")
+var regExpr = regexp.MustCompile("^\\s*\\{\\{\\s*regex:(.*)\\}\\}\\s*$")
+
+func XmlTemplatedMatch(match interface{}, toMatch string) bool {
+	matchString, ok := match.(string)
+	if !ok {
+		return false
+	}
+
+	// parse xml in mock data into dom tree
+	expected := etree.NewDocument()
+	if err := expected.ReadFromString(matchString); err != nil {
+		return false
+	}
+
+	// parse xml in actual request body into dom tree
+	actual := etree.NewDocument()
+	if err := actual.ReadFromString(toMatch); err != nil {
+		return false
+	}
+
+	// tree matching
+	return compareTree(expected.Root(), actual.Root())
+}
+
+func compareTree(expected *etree.Element, actual *etree.Element) bool {
+	// compare constructure
+	// step 1. compare tag name
+	if expected.Tag != actual.Tag {
+		return false
+	}
+	// step 2. compare node content
+	// case 1: leaf
+	if isLeaf(expected) {
+		// compare text content
+		return compareValue(expected.Text(), actual.Text())
+	}
+	// case 2: children element matching
+	actual_children := actual.ChildElements()
+	// for each expected tag
+	for _, match := range expected.ChildElements() {
+		// find one in actual
+		matched := false
+		for i, ele := range actual_children {
+			if compareTree(match, ele) {
+				matched = true
+				// remove matched
+				actual_children = append(actual_children[:i], actual_children[i+1:]...)
+				break
+			}
+		}
+		// all elements in actual data is not matched
+		// or, too many elements in expected data
+		if matched == false {
+			return false
+		}
+	}
+	// too many elements in actual data
+	if len(actual_children) > 0 {
+		return false
+	}
```

Comment:
so this is not needed if it was checked upfront

Reply:
I will show you the logic in pesuocode.
```
expected_children = a list from matcher
actual_children = a list from actual request body, as matching candidates
for (every item in expected_children ) {
    if (item is also in actual_children) {
        item matched, and remove it from matching candidates
    } else {
        item is not found in matching candidates --> [BRANCH 1] matching fails
    }
}
if (any matching candidates remained after we checked all items expected) {
   some unexpected items found in actual request --> [BRANCH 2] matching fails
}
all items are matched --> [BRANCH 3] matching succeed
```

It covers all the cases:
- `len(expected) = len(actual)`
  - all items matched ==> BRANCH 3
  - item in expected does not occurs in actual ==> BRANCH 1
- `len(expected) > len(actual)` ==> BRANCH 1
- `len(expected) < len(actual)` ==> BRANCH 2

We can not do the `if` check for BRANCH 2 before the `for` loop.

Reply:
my bad, this is different from checking the length before the for loop.